### DISCUSSION TITLE
Old user, new Quartz: Object reference not set to an instance of an object

### DISCUSSION THREAD

**Discussion by peteringemann** [quartznet/quartznet#968] [General] [answered]
Hi there,
It's been a couple of years since I used Quartz. Anyways - I just tried to set up a cron trigger for a job using IJob.
This is what I did:

created a class and implemented the IJob and added Debug.Writeline("Hello World") in the Execute-method
then the rest of the code goes as:

```
Dim schedFactory As ISchedulerFactory = New Global.Quartz.Impl.StdSchedulerFactory()
Dim sched As IScheduler = await schedFactory.GetScheduler()
Dim cron As String = "0/20 * * * * ?"
Dim quartzTriggerName As String = "a"
Dim quartzGroupName As String = "b"
Dim quartzJobName As String =  "c"


Dim job As IJobDetail = JobBuilder.Create(GetType(T)) _
    .WithIdentity(quartzJobName, quartzGroupName) _
    .Build()

Dim trigger As ITrigger = TriggerBuilder.Create() _
                    .WithIdentity(quartzTriggerName, quartzGroupName) _
                    .StartAt(DateBuilder.FutureDate(0, IntervalUnit.Second)) _
                    .WithCronSchedule(cron) _
                    .Build()

Await  sched.ScheduleJob(job, trigger)
Await sched.Start()
```

This works but when the job finish I get an exception from line 194 in JobRunShell.cs System.NullReferenceException: 'Object reference not set to an instance of an object.'

I can just ignore the error and then I'll get it at next successful run. My guess is that this is not normal.
Am I just to ignore it or am I doing it wrong?

**Comment by lahma** [comment, marked as answer]
Which version are you using? What are you returning from Execute? If I'd have to guess, I'd say you are returning null instead of task from Execute.

**Reply by peteringemann** [reply]
latest possible from Nuget (3.1.0)
And I am not returning anything. Just writing a debug message

Debug.WriteLine("Hello World!")


...am I supposed to return something? why would I return something? 


update: it works if I return an empty task of string. Why is this necessary?

**Reply by lahma** [reply]
I'm sorry but I'm not fluent with Visual Basic. If the implicit return type is null/undefined, then that causes trouble. Quartz expects your job to return task like the interface dictates `Task Execute(JobExecutionContext)`. So if you don't return a task the .NET runtime will issue an `await` against `null` which will crash.

So always return either a `Task.CompletedTask` or if you have `await`s in your method the return type will become Task.

**Reply by peteringemann** [reply]
In VB.NET you have to define if you're void returns something or doesn't. If it does it's a function, if its not it is a sub. If you create a function that doesn't return anything it's basically just a sub.
When I implement your IJob I have to implement the Execute which IntelliSense marks up as a function. that returns Task. 
Getting the error was just me treating the function as a sub by not returning anything. 

I'll just return Task.CompletedTask - works great! 

Thanks 🙂

**Reply by lahma** [reply]
Thanks for clarifying and great that it's working now!
